### PR TITLE
Add flash support for NXP frdm_ke17z and frdm_ke17z512

### PR DESCRIPTION
**Pull request by ankeXiao**
Tested:
tests/kernel/xip
tests/drivers/flash/common
tests/subsys/fs/fs_api
samples/subsys/fs/littlefs

Report a bug https://github.com/zephyrproject-rtos/zephyr/issues/76321 that littleFs test case does not support non 4096b erase block size.

### CODE REVIEW

**Review by butok — DISMISSED** [review]

**Review by de-nordic — DISMISSED** [review]

File: dts/arm/nxp/nxp_ke17z.dtsi [hunk 1]
```diff
@@ -42,7 +42,7 @@
 			flash0: flash@0 {
 				compatible = "soc-nv-flash";
 				reg = <0 DT_SIZE_K(256)>;
-				erase-block-size = <DT_SIZE_K(2)>;
+				erase-block-size = <DT_SIZE_K(4)>;
```
Comment (Laczen) [inline dts/arm/nxp/nxp_ke17z.dtsi:45]: Why is this changed?

File: dts/arm/nxp/nxp_ke17z512.dtsi [hunk 2]
```diff
@@ -39,7 +39,7 @@
 			flash0: flash@0 {
 				compatible = "soc-nv-flash";
 				reg = <0 DT_SIZE_K(512)>;
-				erase-block-size = <DT_SIZE_K(2)>;
+				erase-block-size = <DT_SIZE_K(4)>;
```
Comment (Laczen) [inline dts/arm/nxp/nxp_ke17z512.dtsi:42]: Same remark, why?

File: dts/arm/nxp/nxp_ke17z.dtsi (hunk 1 above)
Comment (ankeXiao) [reply dts/arm/nxp/nxp_ke17z.dtsi:45]: The littlefs requires a block size of 4K, and 2K is not enough.

Comment (Laczen) [reply dts/arm/nxp/nxp_ke17z.dtsi:45]: This shouldn't be the case. Littlefs should work with a block size of 2kB. Probably there is some error in a littlefs test.

Comment (ankeXiao) [reply dts/arm/nxp/nxp_ke17z.dtsi:45]: I see that all the boards that support littlefs use erase-block-size of 4096, by changing the value of this attribute found that the value of stat.f_frsize in the test code [tests/subsys/fs/littlefs/src/test_lfs_basic.c](https://github.com/zephyrproject-rtos/zephyr/blob/d6afb8fb521844ff8a24d70f1432a428b975a1bd/tests/subsys/fs/littlefs/src/test_lfs_basic.c#L62) is from here, so it should be 4096. Increasing this value has no other effect.

Comment (butok) [reply dts/arm/nxp/nxp_ke17z.dtsi:45]: > I see that all the boards that support littlefs use erase-block-size of 4096

Do you mean that the value was set according to the littlefs SW, not a real HW?

Comment (ankeXiao) [reply dts/arm/nxp/nxp_ke17z.dtsi:45]: So if I want to test the littlefs, this change should be put in the tests/subsys/fs/littlefs boards overlay, right?

Comment (Laczen) [reply dts/arm/nxp/nxp_ke17z.dtsi:45]: You could try to do it in an overlay, but really the test is wrong.

Comment (butok) [reply dts/arm/nxp/nxp_ke17z.dtsi]: > So if I want to test the littlefs, this change should be put in the tests/subsys/fs/littlefs boards overlay, right?

Better to fix the littlefs limitation.

Comment (ankeXiao) [reply dts/arm/nxp/nxp_ke17z.dtsi]: Updated, fixed values replaced with macros, from the dts configuration. Please help to review, thanks!

File: tests/subsys/fs/littlefs/src/testfs_lfs.h [hunk 3]
```diff
@@ -26,6 +27,9 @@ extern struct fs_mount_t testfs_large_mnt;
 #define LARGE_CACHE_SIZE 1024
 #define LARGE_LOOKAHEAD_SIZE 128
 
+#define BLOCK_SIZE DT_PROP(DT_INST(0, soc_nv_flash), erase_block_size)
```
[the remark this answers is not in the document]
Comment (ankeXiao) [reply tests/subsys/fs/littlefs/src/testfs_lfs.h:30]: So can I update this macro only for ke1xz? Like:
image
Or update erase-block-size to 4K in the overlay of ke1xz. The test value in test_lfs_basic function is 4K, and all other platforms are set to 4K.

File: boards/nxp/frdm_ke17z/frdm_ke17z.dts [hunk 4]
```diff
@@ -74,3 +74,31 @@
...
+		boot_partition: partition@0 {
+			label = "mcuboot";
+			reg = <0x0 DT_SIZE_K(44)>;
+		};
+		/* The MCUBoot swap-move algorithm uses the last 2 sectors
+		 * of the primary slot0 for swap status and move.
+		 */
+		slot0_partition: partition@b000 {
+			label = "image-0";
+			reg = <0xb000 (DT_SIZE_K(96) + DT_SIZE_K(8))>;
```
Comment (Laczen) [inline boards/nxp/frdm_ke17z/frdm_ke17z.dts:93]: 4kB extra should suffice. Distribute the 4 kB over slot0 and slot1.

File: boards/nxp/frdm_ke17z512/frdm_ke17z512.dts [hunk 5]
```diff
@@ -86,3 +86,32 @@
...
+			label = "mcuboot";
+			reg = <0x00000000 DT_SIZE_K(64)>;
+			read-only;
+		};
+		/* The MCUBoot swap-move algorithm uses the last 2 sectors
+		 * of the primary slot0 for swap status and move.
+		 */
+		slot0_partition: partition@10000 {
+			label = "image-0";
+			reg = <0x00010000 (DT_SIZE_K(200) + DT_SIZE_K(8))>;
```
Comment (Laczen) [inline boards/nxp/frdm_ke17z512/frdm_ke17z512.dts:106]: Same as above.

**Comment by Laczen** [comment]
@ankeXiao remove the additions for the littlefs test. Create a bug report that the littlefs test does not support non 4096b erase blocks, there is no problem in the flash definition and the correction for littlefs does not belong in this PR.

File: tests/subsys/fs/littlefs/src/testfs_lfs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (butok) [reply tests/subsys/fs/littlefs/src/testfs_lfs.h:30]: > That is problem with all tests scenarios that run on external device.

Guess, we can  extract the sector size information via "chosen->zephyr,flash->erase-block-size".

[the remark this answers is not in the document]
Comment (Laczen) [reply tests/subsys/fs/littlefs/src/testfs_lfs.h:30]: No this also doesn't work, not all devices have this parameter. So I propose to move this problem out of the PR.

File: boards/nxp/frdm_ke17z/frdm_ke17z.dts (hunk 4 above)
Comment (ankeXiao) [reply boards/nxp/frdm_ke17z/frdm_ke17z.dts]: Updated

**Comment by ankeXiao** [comment]
> @ankeXiao remove the additions for the littlefs test. Create a bug report that the littlefs test does not support non 4096b erase blocks, there is no problem in the flash definition and the correction for littlefs does not belong in this PR.

Created issue: https://github.com/zephyrproject-rtos/zephyr/issues/76321, please help to assign it to the owner.

File: tests/subsys/fs/littlefs/src/testfs_lfs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ankeXiao) [reply tests/subsys/fs/littlefs/src/testfs_lfs.h]: Removed

File: boards/nxp/frdm_ke17z/frdm_ke17z.dts [hunk 6]
```diff
@@ -74,3 +74,31 @@
...
+		boot_partition: partition@0 {
+			label = "mcuboot";
+			reg = <0x0 DT_SIZE_K(44)>;
+		};
+		/* The MCUBoot swap-move algorithm uses the last 2 sectors
+		 * of the primary slot0 for swap status and move.
+		 */
+		slot0_partition: partition@b000 {
+			label = "image-0";
+			reg = <0xb000 DT_SIZE_K(2)>;
```
Comment (Laczen) [inline boards/nxp/frdm_ke17z/frdm_ke17z.dts:93]: Something went wrong in the update, the size should be `(DT_SIZE_K(98) + DT_SIZE_K(4))`, and `DT_SIZE_K(98)` for slot1. Addresses need to be recalculated (but you used the correct method :)).

File: boards/nxp/frdm_ke17z512/frdm_ke17z512.dts [hunk 7]
```diff
@@ -86,3 +86,32 @@
...
+			label = "mcuboot";
+			reg = <0x00000000 DT_SIZE_K(64)>;
+			read-only;
+		};
+		/* The MCUBoot swap-move algorithm uses the last 2 sectors
+		 * of the primary slot0 for swap status and move.
+		 */
+		slot0_partition: partition@10000 {
+			label = "image-0";
+			reg = <0x00010000 DT_SIZE_K(2)>;
```
Comment (Laczen) [inline boards/nxp/frdm_ke17z512/frdm_ke17z512.dts:106]: Same as above.

File: tests/subsys/fs/littlefs/boards/frdm_ke17z.overlay [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+ */
+
+&flash0 {
+	partitions {
+		small_partition: partition@25000 {
+			label = "small";
+			reg = <0x25000 DT_SIZE_K(64)>;
+		};
+	};
+};
```
Comment (Laczen) [inline tests/subsys/fs/littlefs/boards/frdm_ke17z.overlay:14]: Remove this file.

File: tests/subsys/fs/littlefs/boards/frdm_ke17z512.overlay [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+&flash0 {
+	partitions {
+		small_partition: partition@44000 {
+			label = "small";
+			reg = <0x00044000 DT_SIZE_K(64)>;
+		};
+
+	};
+};
```
Comment (Laczen) [inline tests/subsys/fs/littlefs/boards/frdm_ke17z512.overlay:15]: Remove this file

File: tests/subsys/fs/littlefs/testcase.yaml [hunk 10]
```diff
@@ -15,7 +15,7 @@ common:
 tests:
   filesystem.littlefs.default:
     timeout: 60
-    platform_allow: nrf54l15pdk/nrf54l15/cpuapp
+    platform_allow: nrf54l15pdk/nrf54l15/cpuapp frdm_ke17z frdm_ke17z512
```
Comment (Laczen) [inline tests/subsys/fs/littlefs/testcase.yaml:18]: Remove this change

File: boards/nxp/frdm_ke17z/frdm_ke17z.dts (hunk 6 above)
Comment (ankeXiao) [reply boards/nxp/frdm_ke17z/frdm_ke17z.dts]: Done

File: tests/subsys/fs/littlefs/testcase.yaml (hunk 10 above)
Comment (ankeXiao) [reply tests/subsys/fs/littlefs/testcase.yaml]: Done

File: tests/subsys/fs/littlefs/boards/frdm_ke17z512.overlay (hunk 9 above)
Comment (ankeXiao) [reply tests/subsys/fs/littlefs/boards/frdm_ke17z512.overlay]: Done

File: tests/subsys/fs/littlefs/boards/frdm_ke17z.overlay (hunk 8 above)
Comment (ankeXiao) [reply tests/subsys/fs/littlefs/boards/frdm_ke17z.overlay]: Done

**Review by Laczen — APPROVED** [review]
LGTM, thanks

**Review by butok — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Review by decsny — APPROVED** [review]
